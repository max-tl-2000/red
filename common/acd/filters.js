/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

import { MainRoleDefinition } from './rolesDefinition';
import { sortOrder, sortRoles, getMinEscalation, getMaxEscalation } from './roles';

const shouldUserBeVisible = (user, teamId, loggedInUserId, loggedInUserMaxEscalation) => {
  if (user.id === loggedInUserId) return true;
  const userTeam = user.teams.find(team => team.id === teamId);

  const userMainRoles = userTeam.mainRoles;
  const minUserEscalation = getMinEscalation(userMainRoles);
  return minUserEscalation <= loggedInUserMaxEscalation;
};

export const dashboardFilterByUser = (loggedInUser, allUsers, selectedUserId) => {
  if (loggedInUser.id === selectedUserId) {
    return {
      users: [selectedUserId],
      teams: loggedInUser.teams.map(team => team.id),
    };
  }

  const selectedUser = allUsers.get(selectedUserId);

  const filteredTeams = selectedUser.teams.filter(selectedUserTeam => {
    const loggedInUserTeam = loggedInUser.teams.find(team => team.id === selectedUserTeam.id);
    if (!loggedInUserTeam) return false;

    const loggedInUserMaxEscalation = getMaxEscalation(loggedInUserTeam.mainRoles);
    const selectedUserMinEscalation = getMinEscalation(selectedUserTeam.mainRoles);
    return loggedInUserMaxEscalation >= selectedUserMinEscalation;
  });

  return {
    users: [selectedUserId],
    teams: filteredTeams.map(team => team.id),
  };
};

export const dashboardFilterByTeam = (loggedInUser, allUsers, selectedTeamId) => {
  const loggedInUserMaxEscalation = getMaxEscalation(loggedInUser.teams.find(team => team.id === selectedTeamId).mainRoles);
  const allTeamUsers = allUsers.filter(user => user.teams.some(team => team.id === selectedTeamId));
  const filteredUsers = allTeamUsers.filter(user => shouldUserBeVisible(user, selectedTeamId, loggedInUser.id, loggedInUserMaxEscalation));
  return {
    users: filteredUsers.toArray().map(user => user.id),
    teams: [selectedTeamId],
  };
};

export const getUserMainRole = roles => {
  const mainRole = roles.sort((role1, role2) => sortRoles(role1, role2, sortOrder.DESC))[0];
  return (mainRole && MainRoleDefinition[mainRole].displayName) || '';
};

export const getUserTitle = user => {
  const allUserRoles = user.teams.reduce((allRoles, team) => [...allRoles, ...team.mainRoles], []);

  return getUserMainRole(allUserRoles);
};
