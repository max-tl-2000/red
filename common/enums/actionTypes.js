/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

export const ActionTypes = {
  QUOTE_UNIT: 'quote_unit',
  PROMOTE_QUOTE: 'promote_quote',
  CREATE_LEASE: 'create_lease',
  PUBLISH_LEASE: 'publish_lease',
  VOID_LEASE: 'void_lease',
  EXECUTE_LEASE: 'execute_lease',
};
