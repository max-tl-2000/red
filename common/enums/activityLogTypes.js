/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

export const ACTIVITY_TYPES = {
  NEW: 'new',
  REMOVE: 'remove',
  REVOKE: 'revoke',
  REJECTED: 'rejected',
  CLOSE: 'close',
  UPDATE: 'update',
  PUBLISH: 'publish',
  TEXT: 'text',
  DUPLICATE: 'duplicate',
  PRINT: 'print',
  EMAIL: 'email',
  ANSWERED: 'answered',
  LISTENED: 'listened',
  FORWARDED: 'forwarded',
  READ: 'read',
  COMPLETED: 'completed',
  SUBMIT: 'submit',
  APPROVE: 'approve',
  DECLINE: 'decline',
  CONFIRM: 'confirm',
  IN_OFFICE_SIGNATURE: 'in office signature',
  SIGN: 'sign',
  COUNTERSIGN: 'countersign',
  EXECUTE: 'execute',
  VIEW: 'view',
  EXPORT: 'export data',
  GUEST_MERGED: 'guest merged',
  MERGE_PARTIES: 'merge parties',
  DONT_MERGE_PARTIES: "don't merge parties",
  ADD_MANUAL_HOLD: 'add hold',
  REMOVE_MANUAL_HOLD: 'remove hold',
  ADD_LEASE_HOLD: 'add lease hold',
  REMOVE_LEASE_HOLD: 'remove lease hold',
  INVENTORY_RESERVED: 'inventory reserved',
  INVENTORY_RELEASED: 'inventory released',
  MOVEOUT: 'moveout',
  MANUAL: 'manual',
  DEACTIVATE: 'deactivate',
  ARCHIVE: 'archive',
  TERMINATED: 'terminated',
  SPAWN: 'spawn',
  DATA_SYNC: 'data sync',
  SET_FLAG: 'set flag',
  OTHER: 'other',
};

export const COMPONENT_TYPES = {
  GUEST: 'guest',
  APPOINTMENT: 'appointment',
  PARTY: 'party',
  PERSON: 'person',
  QUOTE: 'quote',
  EMAIL: 'email',
  SMS: 'sms',
  CALL: 'call',
  LEASINGTEAM: 'leasing team',
  TASK: 'task',
  APPLICATION: 'application',
  CONTACT_EVENT: 'contact event',
  LEASE: 'lease',
  INVENTORY_STATUS: 'inventory status',
  MARKETING_FORM: 'marketing form',
  DIRECT_MESSAGE: 'DM',
};

export const SUB_COMPONENT_TYPES = {
  WAIVER: 'waiver',
  RESCREENING: 'rescreening',
  QUOTE: 'quote',
  RENEWAL_LETTER: 'renewal letter',
  RENEWAL: 'renewal',
  MOVED: 'moved',
  ALL_COMMS_MARKED_AS_READ: 'all comms marked as read',
};
