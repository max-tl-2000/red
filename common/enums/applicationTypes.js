/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

export { FADV_RESPONSE_STATUS } from '../../rentapp/common/screening-constants';

export const ScreeningDecision = {
  APPROVED: 'approved',
  GUARANTOR_REQUIRED: 'Guarantor Required',
  DECLINED: 'declined',
  FURTHER_REVIEW: 'further_review',
  DISPUTED: 'disputed',
  INCOMPLETE: 'incomplete',
  APPROVED_WITH_COND: 'approved_with_cond',
  NO_SCREENING_REQUEST: 'no_screening_request',
  NO_SCREENING_RESPONSE: 'no_screening_response',
  NO_SCREENING_RESPONSE_INTERNATIONAL_ADDRESS: 'no_screening_response_international_address',
  ERROR_ADDRESS_UNPARSABLE: 'error_address_unparsable',
  ERROR_RESPONSE_UNPARSABLE: 'error_response_unparsable',
  ERROR_OTHER: 'error_other',
  SCREENING_IN_PROGRESS: 'screening_in_progress',
  GUARANTOR_DENIED: 'Guarantor Denied',
  COMPILING: 'compiling',
  COMPILING_DELAYED: 'compiling_delayed',
  RESULTS_DELAYED: 'results_delayed',
  ON_HOLD: 'on_hold',
  EXPIRED: 'expired',
  PENDING: 'pending',
  DRAFT: 'draft',
};

export const REVA_SERVICE_STATUS = {
  COMPLETE: 'COMPLETE',
  BLOCKED: 'BLOCKED',
  IN_PROCESS: 'IN_PROCESS',
  INCOMPLETE: 'INCOMPLETE',
};

export const ServiceNames = {
  SSN: 'SSN',
  CREDIT: 'Credit',
  EVICTION: 'Eviction',
  SKIPWATCH: 'SkipWatch',
  CRIMINAL: 'Criminal',
  SEXOFFENDER: 'SexOffender',
  COLLECTIONS: 'Collections',
  CRIMINAL_OFFLINE: 'Criminal Offline',
  SEX_OFFENDER_OFFLINE: 'Sex Offender Offline',
  GLOBAL_SANCTIONS_OFFLINE: 'Global Sanctions Offline',
  NEAR_INSTANT: 'NearInstant',
};

export const GroupServiceNames = {
  CREDIT: 'credit',
  CRIMINAL: 'criminal',
};

export const ApplicationSections = {
  incomeSourcesSection: 'incomeSourcesSection',
  addressHistorySection: 'addressHistorySection',
  childrenSection: 'childrenSection',
  disclosuresSection: 'disclosuresSection',
  petsSection: 'petsSection',
  privateDocumentsSection: 'privateDocumentsSection',
  rentersInsuranceSection: 'rentersInsuranceSection',
  sharedDocumentsSection: 'sharedDocumentsSection',
  vehiclesSection: 'vehiclesSection',
};

export const ApplicationSettingsTypes = {
  holdDepositWithoutUnit: 'holdDepositWithoutUnit',
};

export const HoldDepositApplicationSettingsValues = {
  REQUIRED_BY_FIRST_RESIDENT_APPLICANT: 'requiredByFirstResidentApplicant',
  OPTIONAL: 'optional',
  HIDDEN: 'hidden',
};

export const ApplicationSettingsValues = {
  REQUIRED: 'required',
  OPTIONAL: 'optional',
  HIDDEN: 'hidden',
};

export const SharedSections = [
  ApplicationSections.petsSection,
  ApplicationSections.childrenSection,
  ApplicationSections.vehiclesSection,
  ApplicationSections.sharedDocumentsSection,
];

export const BlockedReasons = {
  DISPUTE: 'dispute',
  ADDRESS: 'address',
  EXPIRED: 'expired',
  UNKNOWN: 'unknown',
  CREDIT_FREEZE: 'creditFreeze',
};

export const approvedScreening = ['approved'];
export const conditionalScreening = ['approved_with_cond', 'further_review', 'disputed', 'guarantor required'];
export const deniedScreening = ['declined'];
