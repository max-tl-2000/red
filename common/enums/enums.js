/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

export const LOG_MESSAGE_MAX_LENGTH = 256;
export const NO_IMAGE_RESIZE = -1;
export const GLOBAL_SEARCH_QUERY_MAX_LENGTH = 255;
export const COMPANY_NAME_MAX_LENGTH = 255;

export const PhoneOwnerType = {
  TEAM: 'team',
  TEAM_MEMBER: 'team_member',
  USER: 'user',
  PROGRAM: 'program',
};

export const YardiResidentImportFiles = {
  ResTenants: 'ResTenants',
  ResRoommates: 'ResRoommates',
};

export const YardiUnitsImportFiles = {
  CommUnits: 'CommUnits',
};

export const RmsImportError = {
  FILE_NOT_RECOGNIZED_ERROR: 'FILE_NOT_RECOGNIZED',
  NO_FILES_ADDED_ERROR: 'NO_FILES_ADDED',
  PARSING_FAILED_ERROR: 'PARSING_FAILED',
  EMPTY_FILE_ERROR: 'EMPTY_FILE',
  FILE_NOT_FOUND_ERROR: 'FILE_NOT_FOUND',
  S3_UPLOAD_FAILED_ERROR: 'S3_UPLOAD_FAILED',
  DATABASE_SAVE_FAILED_ERROR: 'DATABASE_SAVE_FAILED',
  PROPERTY_NOT_FOUND_IN_DB_ERROR: 'PROPERTY_NOT_FOUND_IN_DB',
  INVENTORY_NOT_FOUND_IN_DB_ERROR: 'INVENTORY_NOT_FOUND_IN_DB',
  LEASE_TERM_MISMATCH_ERROR: 'LEASE_TERM_MISMATCH',
  PRICES_NOT_FOUND_IN_DB_ERROR: 'PRICES_NOT_FOUND_IN_DB',
};

export const RmsPricingEvents = {
  EXTERNAL_RMS_IMPORT: 'EXTERNAL_RMS_IMPORT',
  REVA_IMPORT: 'REVA_IMPORT',
  INVENTORY_STATE_CHANGE: 'INVENTORY_STATE_CHANGE',
};

export const ImportMappersEntityTypes = {
  UnitStatusMapper: 'inventory',
  RentableItemsMapper: 'inventory',
  UnitAmenitiesMapper: 'unitAmenities',
  ProspectsMapper: 'prospects',
  HistoricalCommunicationMapper: 'historicalCommunication',
  MriUnitMapper: 'inventory',
  MriUnitAmenitiesMapper: 'mriUnitAmenities',
  MriProperties: 'mriProperties',
  MriRentableItems: 'mriRentableItems',
  PartiesMapper: 'parties',
  PartyMembersMapper: 'partyMembers',
  CohortFile: 'cohortFile',
  CommUnitsMapper: 'commonUnits',
};

export const ResetPasswordTypes = {
  ADMIN: 'admin',
  USER_DEFAULT: 'user default',
  REVA_ADMIN: 'reva admin',
  SFTP: 'SFTP',
  LRO: 'LRO',
};

export const AppointmentEmailType = {
  CREATE: 'create',
  UPDATE: 'update',
  CANCEL: 'cancel',
};

export const LeaseEmailType = {
  SENT: 'sent',
  VOID: 'void',
  EXECUTE: 'execute',
};

export const AppointmentContextTypes = {
  MOST_RECENT: 'mostRecent',
  UPCOMING: 'upcoming',
};

export const StaticAssetType = {
  SWATCHES: 'swatches',
  MASKS: 'masks',
};

export const AssetTheme = {
  LIGHT: 'light',
  DARK: 'dark',
};

export const RetractedPostReasons = {
  INACCURATE_OR_MISLEADING_INFORMATION: 'INACCURATE_OR_MISLEADING_INFORMATION',
  NO_LONGER_APPLICABLE_OR_VALID: 'NO_LONGER_APPLICABLE_OR_VALID',
  MISSING_OR_INCOMPLETE_INFORMATION: 'MISSING_OR_INCOMPLETE_INFORMATION',
  SENT_TO_THE_WRONG_GROUP_OF_PEOPLE: 'SENT_TO_THE_WRONG_GROUP_OF_PEOPLE',
};

export const AssetThemeTypes = [AssetTheme.LIGHT, AssetTheme.DARK];

export const PaymentMethodCallbackResult = {
  SUCCESS: 'success',
  CANCEL: 'cancel',
};

export const LeaseProviderName = {
  FADV: 'FADV',
  BLUEMOON: 'BLUEMOON',
};

export const telephonyDisconnectedReasons = {
  USER_REFUSED_MIC_ACCESS: 'user refused mic access',
  NO_INTERNET_CONNECTION: 'no internet connection',
  OTHER: 'other',
};
