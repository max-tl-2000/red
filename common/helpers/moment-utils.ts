/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

import moment, { Moment } from 'moment-timezone'; // eslint-disable-line red/no-moment
import * as baseMoment from 'moment'; // eslint-disable-line red/no-moment
import { extendMoment } from 'moment-range';
import nullish from './nullish';
import { isString } from './type-of';
import { isNode } from './globals';

export type MomentType = Moment | string | Date;

interface IIsSameDayOptions {
  timezone?: string;
}

interface ILogWarningsArgs {
  msg?: any[];
}

interface IToMomentArgs {
  parseFormat?: string;
  timezone?: string;
  strict?: boolean;
}

interface IFormatMomentArgs {
  format?: string;
  timezone?: string;
  includeZone?: boolean;
  includeZoneStrict?: boolean;
}

interface INowArgs {
  timezone?: string;
}

interface IParseASInTimezoneArgs {
  format?: string;
  timezone?: string;
}

const runningInNode = isNode();

const momentFn = moment;
const { duration, max: momentMax, now: momentNow } = momentFn;

export const DATE_ISO_FORMAT = 'YYYY-MM-DD';

// this format is not correct the UTC formats with `Z` don't include +z
export const DATE_TIME_ISO_FORMAT = 'YYYY-MM-DDTHH:mm:ss+z';

export const findLocalTimezone = (): string => momentFn.tz.guess();
export const isValidTimezone = (tz: string): boolean => !!momentFn.tz.zone(tz);

let momentUtilsFixedNow;

const logWarnings = ({ msg = [] }: ILogWarningsArgs = {}) => {
  const isProduction = process.env.NODE_ENV === 'production';

  if (isProduction || !msg || msg.length === 0) {
    return;
  }

  const isDevelopment = process.env.NODE_ENV === 'development';

  console.warn(...msg);

  if (isDevelopment && runningInNode) {
    const err = new Error();
    const LINES_TO_SKIP = 2;
    // we remove the 2 first lines which are meaningless for us
    // and also the lines that contains stack info from teh asyncGenerator babel transform
    const stackWithMeaningFulInfo = (err.stack || '')
      .split(/\n/)
      .slice(LINES_TO_SKIP)
      .filter(line => !line.match(/asyncToGenerator\.js/))
      .join('\n');

    console.warn(stackWithMeaningFulInfo);
  }
};

const checkIfDateIsISO = (date: MomentType): void => {
  const isProduction = process.env.NODE_ENV === 'production';

  if (isProduction || !isString(date)) {
    return;
  }

  date = date as string; // cast needed to make typescript happy

  const ISO_8601_REGEX_FULL = /^\d{4}-\d\d-\d\dT\d\d:\d\d:\d\d(\.\d+)?(([+-]\d\d:\d\d)|Z)?$/i;
  const ISO_8601_REGEX_PERMISSIVE = /^\d{4}(-\d\d(-\d\d(T\d\d:\d\d(:\d\d)?(\.\d+)?(([+-]\d\d:\d\d)|Z)?)?)?)?$/i;

  if (date.match(ISO_8601_REGEX_FULL)) {
    return;
  }

  if (date.match(ISO_8601_REGEX_PERMISSIVE)) {
    // if it matches the permissive then we just inform it failed the full ISO validation
    logWarnings({ msg: ['>> `date` parameter is not in full ISO format', date] });
    return;
  }

  logWarnings({ msg: ['>> date parameter is not in any ISO format', date] });
};

export const toMoment = (date: MomentType, { parseFormat, timezone, strict = true }: IToMomentArgs = { strict: true }): Moment => {
  if (nullish(date) || date === '') {
    logWarnings({
      msg: ['>> date parameter is nullish or empty', date, '\n', '>> this will produce an invalid moment that will be formatted as empty string'],
    });

    const m = momentFn(null!);
    m.format = () => '';
    return m; // just return an invalid moment object
  }

  if (parseFormat !== undefined && typeof parseFormat !== 'string') {
    throw new TypeError('`parseFormat` must be a string');
  }

  checkIfDateIsISO(date);

  if (timezone !== undefined && !isString(timezone)) {
    throw new TypeError('timezone must be a string');
  }

  let dateAsMoment = parseFormat ? momentFn(date, parseFormat, strict) : momentFn(date);

  if (timezone) {
    dateAsMoment = dateAsMoment.tz(timezone);
  }

  return dateAsMoment;
};

export const formatMoment = (
  momentObj: Moment,
  { format, timezone, includeZone = true, includeZoneStrict = false }: IFormatMomentArgs = { includeZone: true, includeZoneStrict: false },
): string => {
  if (!timezone) {
    logWarnings({ msg: ['[formatMoment] called without timezone', momentObj] });
  }

  const m = momentObj ? toMoment(momentObj, { timezone }) : undefined;

  if (!m || !m.isValid()) {
    logWarnings({ msg: ['[formatMoment] attempt to format an invalid moment', momentObj] });
    return '';
  }

  const useAbbr = timezone && includeZone && timezone !== findLocalTimezone();

  const zone = useAbbr || includeZoneStrict ? ` ${m.zoneAbbr()}` : '';
  return `${m.format(format)}${zone}`;
};

export const now = ({ timezone = findLocalTimezone() }: INowArgs = { timezone: findLocalTimezone() }) => {
  if (momentUtilsFixedNow) {
    return momentUtilsFixedNow;
  }

  if (timezone !== undefined && typeof timezone !== 'string') {
    throw new TypeError('timezone must be a string');
  }

  let m = momentFn();
  if (timezone) {
    m = m.tz(timezone);
  }
  return m;
};

export const useFixedDateTimeForMomentUtilsNow = dateTime => {
  momentUtilsFixedNow = dateTime;

  return () => {
    momentUtilsFixedNow = undefined;
  };
};

export const parseAsInTimezone = (str: string, { format, timezone }: IParseASInTimezoneArgs = {}): Moment => {
  const args = [str];

  if (format) {
    args.push(format);
  }

  if (timezone) {
    args.push(timezone);
  } else {
    logWarnings({ msg: ['[parseAsInTimezone] error: missing `timezone`. In future versions this will throw an error'] });
  }

  // @ts-ignore: spread will work just fine in this case
  return momentFn.tz(...args);
};

export const isMoment = (m: any): boolean => momentFn.isMoment(m);

export const isSameDay = (date1, date2, { timezone }: IIsSameDayOptions = {}): boolean => {
  if (!timezone) {
    logWarnings({ msg: ['[isSameDay] called without timezone'] });
  }

  date1 = toMoment(date1, { timezone });
  date2 = toMoment(date2, { timezone });

  return !date1.isValid() || !date2.isValid() ? false : date1.isSame(date2, 'day');
};

export const isValidMoment = (m: any): boolean => m && isMoment(m) && m.isValid();

export const setDefaultTZ = (tz: string) => momentFn.tz.setDefault(tz);

export const getDurationBetweenMoments = (startDate: MomentType, endDate: MomentType): string => {
  const millisecondDifference = toMoment(endDate).diff(startDate);
  return moment.utc(millisecondDifference).format('mm:ss');
};

export const rangeMoment = extendMoment(baseMoment);

export const getTimezoneNames = (): Array<string> => momentFn.tz.names();

export const stringRepresentsADateWithoutTime = date => {
  if (typeof date !== 'string') {
    return false;
  }
  const POSIBLE_DATE = /^\d{4}-\d{1,2}-\d{1,2}$/g;
  return !!date.match(POSIBLE_DATE);
};

export const isWithinLastNDays = ({ date, days, timezone }) => toMoment(date, { timezone }).isSameOrAfter(now().add(days, 'days'), 'day');

export { duration, momentMax, momentNow };

export const today = ({ timezone = findLocalTimezone() }: INowArgs = { timezone: findLocalTimezone() }) =>
  now({ timezone }).set({ hour: 0, minute: 0, second: 0, millisecond: 0 });
