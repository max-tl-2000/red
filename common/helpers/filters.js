/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

import { DALTypes } from '../enums/DALTypes';
import { now } from './moment-utils';

export const normalizeFilters = (filters = {}) => {
  const newFilters = {
    propertyIds: filters.propertyIds || [],
    numBedrooms: filters.numBedrooms || [],
    moveInDate: filters.moveInDate || { min: null, max: null },
    marketRent: filters.marketRent || { min: null, max: null },
    lifestyles: filters.lifestyles || [],
    amenities: filters.amenities || [],
    floor: filters.floor || [],
    // unitName: filters.unitName, to get by unitName%. this uses "LIKE"
    // withoutLimit: filters.withoutLimit || false, to get results without the default limit (20 items)
    // query: filters.query || '', to get results using a query, this use globalSearchVector, this will avoid unitName filter
    // inventoryStates: filters.inventoryStates || [], to get units by state like ['model', 'vacantReady']
  };
  if (filters.favoriteUnits) {
    newFilters.favoriteUnits = filters.favoriteUnits;
  }

  return newFilters;
};

export const addPropertyToFilters = (propertyId, filters = {}) => {
  if (!propertyId) return filters;

  const propertyIds = filters.propertyIds || [];
  return {
    ...filters,
    propertyIds: Array.from(new Set([...propertyIds, propertyId])),
  };
};

const addMonthsToDateInTimezone = timezone => monthsToAdd => now({ timezone }).startOf('day').add(monthsToAdd, 'months');

export const createMoveInFilter = (moveInTimeId, { timezone } = {}) => {
  const addMonthsToDate = addMonthsToDateInTimezone(timezone);

  switch (moveInTimeId) {
    case DALTypes.QualificationQuestions.MoveInTime.NEXT_4_WEEKS:
      return { min: addMonthsToDate(0), max: addMonthsToDate(1) };
    case DALTypes.QualificationQuestions.MoveInTime.NEXT_2_MONTHS:
      return { min: addMonthsToDate(1), max: addMonthsToDate(2) };
    case DALTypes.QualificationQuestions.MoveInTime.NEXT_4_MONTHS:
      return { min: addMonthsToDate(2), max: addMonthsToDate(4) };
    case DALTypes.QualificationQuestions.MoveInTime.BEYOND_4_MONTHS:
      return { min: addMonthsToDate(4), max: '' };
    case DALTypes.QualificationQuestions.MoveInTime.I_DONT_KNOW:
    case null:
      return { min: '', max: '' };
    default:
      throw new Error('Invalid key for Move in time filter.');
  }
};

export const getMoveInTimeQuestionFromDate = (moveInDate, { timezone, defaultAnswer = DALTypes.QualificationQuestions.MoveInTime.I_DONT_KNOW } = {}) => {
  const addMonthsToDate = addMonthsToDateInTimezone(timezone);
  if (!moveInDate || moveInDate.isBefore(addMonthsToDateInTimezone(0))) return defaultAnswer;

  if (moveInDate.isAfter(addMonthsToDate(4))) return DALTypes.QualificationQuestions.MoveInTime.BEYOND_4_MONTHS;
  if (moveInDate.isAfter(addMonthsToDate(2))) return DALTypes.QualificationQuestions.MoveInTime.NEXT_4_MONTHS;
  if (moveInDate.isAfter(addMonthsToDate(1))) return DALTypes.QualificationQuestions.MoveInTime.NEXT_2_MONTHS;

  return DALTypes.QualificationQuestions.MoveInTime.NEXT_4_WEEKS;
};

export const updateNumBedroomsToFilters = (numBedrooms, filters) => {
  if (!numBedrooms) return filters;
  return {
    ...filters,
    numBedrooms: Array.from(new Set([...numBedrooms])),
  };
};
