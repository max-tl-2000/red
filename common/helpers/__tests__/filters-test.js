/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

import sinon from 'sinon';
import { DALTypes } from '../../enums/DALTypes';
import { createMoveInFilter } from '../filters';
import { toMoment } from '../moment-utils';

describe('execute createMoveInFilter function', () => {
  let clock;
  const timezoneArgs = { timezone: 'America/Los_Angeles' };

  beforeEach(() => {
    const m = toMoment('2018-02-21T14:30:00Z', timezoneArgs);
    clock = sinon.useFakeTimers(m.valueOf());
  });

  afterEach(() => {
    clock.restore();
  });

  describe('given a NEXT_4_WEEKS moveInTime', () => {
    it('should return a valid filter', () => {
      const result = createMoveInFilter(DALTypes.QualificationQuestions.MoveInTime.NEXT_4_WEEKS, timezoneArgs);
      expect(result).toMatchSnapshot();
    });
  });

  describe('given a NEXT_2_MONTHS moveInTime', () => {
    it('should return a valid filter', () => {
      const result = createMoveInFilter(DALTypes.QualificationQuestions.MoveInTime.NEXT_2_MONTHS, timezoneArgs);
      expect(result).toMatchSnapshot();
    });
  });

  describe('given a NEXT_4_MONTHS moveInTime', () => {
    it('should return a valid filter', () => {
      const result = createMoveInFilter(DALTypes.QualificationQuestions.MoveInTime.NEXT_4_MONTHS, timezoneArgs);
      expect(result).toMatchSnapshot();
    });
  });

  describe('given a BEYOND_4_MONTHS moveInTime', () => {
    it('should return a valid filter', () => {
      const result = createMoveInFilter(DALTypes.QualificationQuestions.MoveInTime.BEYOND_4_MONTHS, timezoneArgs);
      expect(result).toMatchSnapshot();
    });
  });

  describe('given a I_DONT_KNOW moveInTime', () => {
    it('should return a valid filter', () => {
      const result = createMoveInFilter(DALTypes.QualificationQuestions.MoveInTime.I_DONT_KNOW, timezoneArgs);
      expect(result).toEqual({ min: '', max: '' });
    });
  });

  describe('given a null moveInTime', () => {
    it('should a empty filter', () => {
      const result = createMoveInFilter(null);
      expect(result).toEqual({ min: '', max: '' });
    });
  });
});
