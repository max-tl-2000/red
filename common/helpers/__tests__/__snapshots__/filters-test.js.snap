// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`execute createMoveInFilter function given a BEYOND_4_MONTHS moveInTime should return a valid filter 1`] = `
Object {
  "max": "",
  "min": "2018-06-21T07:00:00.000Z",
}
`;

exports[`execute createMoveInFilter function given a NEXT_2_MONTHS moveInTime should return a valid filter 1`] = `
Object {
  "max": "2018-04-21T07:00:00.000Z",
  "min": "2018-03-21T07:00:00.000Z",
}
`;

exports[`execute createMoveInFilter function given a NEXT_4_MONTHS moveInTime should return a valid filter 1`] = `
Object {
  "max": "2018-06-21T07:00:00.000Z",
  "min": "2018-04-21T07:00:00.000Z",
}
`;

exports[`execute createMoveInFilter function given a NEXT_4_WEEKS moveInTime should return a valid filter 1`] = `
Object {
  "max": "2018-03-21T07:00:00.000Z",
  "min": "2018-02-21T08:00:00.000Z",
}
`;
